src/sys/unix/waker.rs

```diff
@@ -21,16 +20,18 @@ mod eventfd {
 
     impl Waker {
         pub fn new(selector: &Selector, token: Token) -> io::Result<Waker> {
-            let fd = syscall!(eventfd(0, libc::EFD_CLOEXEC | libc::EFD_NONBLOCK))?;
-
-            selector.register(fd, token, Interests::READABLE)?;
-            Ok(Waker {
-                fd: unsafe { File::from_raw_fd(fd) },
+            syscall!(eventfd(0, libc::EFD_CLOEXEC | libc::EFD_NONBLOCK)).and_then(|fd| {
+                // Turn the file descriptor into a file first so we're ensured
+                // it's closed when dropped, e.g. when register below fails.
+                let file = unsafe { File::from_raw_fd(fd) };
```

Comment:
Do we want to define our own type that takes an `Fd` and calls `close` on drop? That might be a bit more reliable?

Reply:
On Unix that is exactly what `File` is and since it's not exposed anywhere I would argue that the extra code doesn't outweigh the benefits.

Reply:
The primary danger, though low probability, is the drop impl starts adding file specific syscalls. Up to you though, I approved.

Reply:
I guess that could be the case, let's tackle that when it comes up.